CONTRIBUTING.md

```diff
@@ -0,0 +1,147 @@
+For example:
+
+```python
+def stencil(...):
+    """This is a short description that fits on one line.
+
+    Here is a longer explanation of the assumptions and why this exists.
+
+    TODOC: I do not fully understand why ke_c needs to be updated here.
+
+    Args:
+        uc: x-velocity on C-grid (inout)
+        vc: y-velocity on C-grid (inout)
+        vort_c: Vorticity on C-grid (inout)
+        ke_c: kinetic energy on C-grid (inout)
+        v: y-velocity on D-grid (inout)
+        u: x-velocity on D-grid (inout)
+        dt2: timestep (in)
+
+    """
+```
+
+
+### Python functions
+These should mostly be lightweight workflow wrappers calling gt4py stencils, though currently exceptions exist where Python code does computations on data fields.
+
+Original convention is:
+```python
+def compute(var1, var2, var3, param1, param2, param3):
+```
+
+Order of arguments does not actually matter, but generally follows the convention of listing 3d fields first, followed by parameters, as is required by gt4py stencil functions.
+
+New convention: make use of `fv3core/utils/typing.py` to specify fields, also type-hint any function
+outputs.
+
+For example:
+```python
+def compute(var1: FloatField, var2:IntField, var3: BoolField,
+            param1: float_type, param2: int_type, param3: bool_type):
+```
+
+Another example
+```python
+def make_storage_from_shape(shape, origin, dtype, init=True):
+```
+
+Turns into
+```python
+    def make_storage_from_shape(
+        shape: Tuple[int, int, int],
+        origin: Tuple[int, int, int],
+        *,
+        dtype: DTypes = np.float64,
+        init: bool = True,
+    ) -> Field:
+```
+
+- We will prioritize adding typing to methods that are used by other modules.
+  Not every internal method needs this level of specification.
+- Internal functions that are likely to be inlined into a larger stencil do not need this if it will just be removed in the near-term.
+
+### GT4Py stencils
+FV3core defines a custom decorator `fv3core.gtstencil` defined in `decorators.py` that it uses to define stencils.
+This eventually calls `gt4py.gtscript.stencil`, but sets default external arguments such as `backend`, and `rebuild` and provides the global namelist to the stencils as `namelist`.
+The type of each input of a stencil requires a type and the first version of the model used a shorthand 'sd' (storage data) to indicate a 3D gt4py storage, such as
+
+```python
+@gtstencil()
+def pt_adjust(pkz:sd, dp1: sd, q_con: sd, pt: sd):
+    with computation(PARALLEL), interval(...):
+```
+
+Note that `fv3core.gtstencil` can be manually called on an undecorated stencil, but this is currently in general discouraged except when used internally.
+
+In the refactoring of the dycore, we are using lower dimensional storages and different item types, so `sd` is insufficient to type these.
```

Comment:
I would suggest removing this guideline, since contributed code would be by definition internal code.

```suggestion
In the refactoring of the dycore, we are using lower dimensional storages and different item types, so `sd` is insufficient to type these.
```

Reply:
I think it's meant to call out that there are two ways to call stencils, using the explicit decorator @, or creating the decorated stencil inline. These two styles can be confusing to keep track of.